File: sdks/java/fm-sdk/src/test/java/fm/TokenAuthenticationTest.java
package fm;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * Connecting with a token that is already held, rather than with credentials.
 *
 * <p>This is how every robot connects: the container holds the participant's
 * JWT and passes it as {@code -C}, so there is no account, email or password to
 * present. {@link HttpFlexemarkets} must therefore <em>refresh</em> the token —
 * which validates it and returns the account and person behind it — rather than
 * sign in.
 *
 * <p>The SDK signed in regardless, POSTing {@code /tokens} with the blank
 * username and password that {@code loadCredential} leaves behind, and the
 * server rightly answered 401. No robot could connect. It was invisible to
 * every other test because it needs a server to refuse you.
 *
 * <p>It is also the second time: fm-lib-net carries the same branch with a
 * comment recording that an earlier rewrite dropped it. Hence this test, which
 * stands up a real HTTP server on a loopback port and asserts the request the
 * SDK makes — so the branch cannot be dropped a third time in silence.
 */
class TokenAuthenticationTest {

    private static final String TOKEN =
            "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJkZXZAZGV2In0.c2lnbmF0dXJl";

    private HttpServer server;
    private final List<String> requests = new ArrayList<>();

    @BeforeEach
    void startServer() throws IOException {
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);

        server.createContext("/api/tokens", exchange -> {
            requests.add(exchange.getRequestMethod() + " " + exchange.getRequestURI().getPath()
                         + " auth=" + exchange.getRequestHeaders().getFirst("Authorization"));
            // A sign-in attempt with no credentials is refused, exactly as the
            // real server refuses it. Only the refresh path should succeed.
            if (!exchange.getRequestURI().getPath().endsWith("/refresh")) {
                respond(exchange, 401, "{\"message\":\"unauthorized\"}");
                return;
            }
            respond(exchange, 200, """
                {"token":"%s",
                 "person":{"id":7,"accountId":1,"email":"dev@dev","firstName":"Dev","lastName":"User"},
                 "account":{"id":1,"name":"dev"}}
                """.formatted(TOKEN));
        });

        server.createContext("/api/marketplaces/1", exchange -> {
            requests.add(exchange.getRequestMethod() + " " + exchange.getRequestURI().getPath());
            respond(exchange, 200, "{\"id\":1,\"name\":\"Test\",\"markets\":[]}");
        });

        server.createContext("/api", exchange -> {
            requests.add(exchange.getRequestMethod() + " " + exchange.getRequestURI().getPath());
            respond(exchange, 200, "{\"_links\":{}}");
        });

        server.start();
    }

    @AfterEach
    void stopServer() {
        if (server != null) {
            server.stop(0);
        }
    }

    private static void respond(HttpExchange exchange, int status, String body) throws IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().add("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, bytes.length);
        try (OutputStream out = exchange.getResponseBody()) {
            out.write(bytes);
        }
    }

    private String endpoint() {
        return "http://127.0.0.1:" + server.getAddress().getPort() + "/api/marketplaces/1";
    }

    @Test
    void connectingWithATokenRefreshesItRatherThanSigningIn() throws Exception {
        try (Flexemarkets fm = Flexemarkets.connect(TOKEN, endpoint(), "token-test")) {
            assertThat(fm.userId()).isEqualTo(7L);
            assertThat(fm.accountName()).isEqualTo("dev");
        }

        assertThat(requests)
                .as("a held token is refreshed, never exchanged for another")
                .anySatisfy(r -> assertThat(r).contains("GET /api/tokens/refresh"))
                .noneSatisfy(r -> assertThat(r).isEqualTo("POST /api/tokens auth=null"));
    }

    @Test
    void theTokenIsPresentedAsABearer() throws Exception {
        try (Flexemarkets fm = Flexemarkets.connect(TOKEN, endpoint(), "token-test")) {
            assertThat(fm.userId()).isEqualTo(7L);
        }

        assertThat(requests).anySatisfy(r -> {
            assertThat(r).contains("/api/tokens/refresh");
            assertThat(r).contains("auth=Bearer " + TOKEN);
        });
    }

    /**
     * The failure this replaced. A POST to /tokens carrying the blanks that
     * loadCredential leaves is what the server answered 401 to, and the only
     * symptom was a robot that would not start.
     */
    @Test
    void theSignInEndpointIsNotPostedToAtAll() throws Exception {
        try (Flexemarkets fm = Flexemarkets.connect(TOKEN, endpoint(), "token-test")) {
            assertThat(fm.userId()).isEqualTo(7L);
        }

        assertThat(requests)
                .noneSatisfy(r -> assertThat(r).startsWith("POST /api/tokens auth"));
    }
}

File: sdks/typescript/test/token-authentication.test.ts
/**
 * Connecting with a token rather than a password.
 *
 * This never worked. The SDK POSTed `/tokens` with the bearer header *and* a
 * body of `{"username": "|", "password": ""}`, and fm-server answers that with
 * 400 MESSAGE_NOT_READABLE — verified against a running server, where the same
 * POST with a real password returns 200. So every token connection failed, in
 * this SDK and the Python one, from the day each was written.
 *
 * A caller holding a token has no account, email or password to present. The
 * route that exists for them is `GET /tokens/refresh`, which validates the
 * token and returns the account and person behind it. The Java SDK has always
 * used it, and records that fm-lib-net carried the same branch and an earlier
 * rewrite dropped it — this is the third occurrence.
 *
 * Asserted against a loopback server rather than a stubbed fetch, because a
 * stub is precisely what hid it: a mock answers whatever the test tells it to,
 * and the server's 400 never appears.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import type { AddressInfo } from "node:net";

import { Flexemarkets } from "../src/client.ts";

const TOKEN = "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJkZXZAZGV2In0.c2lnbmF0dXJl";

async function withServer(
  run: (base: string, requests: string[]) => Promise<void>,
): Promise<void> {
  const requests: string[] = [];

  const server = http.createServer((req, res) => {
    requests.push(`${req.method} ${req.url}`);

    if (req.method === "GET" && req.url === "/api/tokens/refresh") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({
        token: TOKEN,
        person: { id: 7, accountId: 1, email: "dev@dev" },
        account: { id: 1, name: "dev" },
      }));
      return;
    }
    if (req.method === "GET" && req.url === "/api") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ _links: {} }));
      return;
    }
    if (req.method === "POST") {
      // What fm-server actually answers a token POST with blanks.
      res.writeHead(400, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ error: "MESSAGE_NOT_READABLE", path: req.url }));
      return;
    }
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end("[]");
  });

  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", () => resolve()));
  const base = `http://127.0.0.1:${(server.address() as AddressInfo).port}/api`;
  try {
    await run(base, requests);
  } finally {
    server.close();
  }
}

test("a token connects through the refresh route", async () => {
  await withServer(async (base, requests) => {
    const fm = await Flexemarkets.connect(TOKEN, `${base}/marketplaces/1`, "token-test");
    try {
      assert.equal(fm.user.id, 7);
      assert.equal(fm.account.name, "dev");
    } finally {
      await fm.close();
    }

    assert.ok(requests.includes("GET /api/tokens/refresh"), requests.join(", "));
  });
});

test("a token never posts to /tokens", async () => {
  // The POST is the defect. A server that refuses it must not be reached.
  await withServer(async (base, requests) => {
    const fm = await Flexemarkets.connect(TOKEN, `${base}/marketplaces/1`, "token-test");
    await fm.close();

    assert.ok(
      !requests.some((r) => r.startsWith("POST /api/tokens")),
      requests.join(", "),
    );
  });
});
